### PR TITLE
Subject by Industry performance improvements

### PR DESCRIPTION
**Pull request by rmbielby**
Reorders and adapts the code in Subject by Industry to reduce unnecessary and duplicated processing. Also fixed a couple of bugs along the way.

### CODE REVIEW

**Review by chfoster — APPROVED** [review]

**Comment by cjrace** [comment]
Noticed a few errors / warnings in the console when running this locally. The NA ones are fine if we're happy with that behaviour, we should just mute them to avoid cluttering the logs.

On loading the tab:
> Warning: Problem while computing `SECTIONNAME = ifelse(...)`. i NAs introduced by coercion

Click median earnings (while 'Sex' or 'Region' or 'Prior attainment' or 'Subject' is selected):

image

> Warning: Error in subset: object 'crosstabs_data_table' not found
  116: subset [C:\R_projects\leo-graduate-industry-dashboard\R\subject_industry.R#1397]
  115: crosstabs [C:\R_projects\leo-graduate-industry-dashboard\R\subject_industry.R#1397]
  114: <reactive:reactiveSubjIndTable> [C:\R_projects\leo-graduate-industry-dashboard/server.R#165]
   98: reactiveSubjIndTable
   97: ::
htmlwidgets
shinyRenderWidget [C:\R_projects\leo-graduate-industry-dashboard/server.R#168]
   96: func
   83: renderFunc
   82: output$crosstab
    1: runApp

**Comment by rmbielby** [comment]
Good spot, will see what's caused that error, looks like an old variable reference.

File: R/subject_industry.R [hunk 1]
```diff
@@ -1,7 +1,962 @@
...
+        x <- paste(data$region[1:nrow(data) - 1], collapse = ", ")
+        y <- paste(" and ", data$region[nrow(data)], sep = "")
+        paste(x, y)
+      } else if (length(data) == 1) {
+        paste(data$region[1])
+      }
+    }
+
+
+    print(length(uniqueregions))
```
Comment (cjrace) [inline R/subject_industry.R]: Don't think this print() is needed - were you using it while developing?

File: R/subject_industry.R [hunk 2]
```diff
@@ -1,7 +1,962 @@
...
+                      <b>", uniqueregions[2], "</b> was the most common industry for those living in ", textprod(data2), ", <b>",
+        uniqueregions[3], "</b> was the most common industry for those living in ", textprod(data3), ", <b>",
+        uniqueregions[4], "</b> was the most common industry for those living in ", textprod(data4), ", <b>",
+        uniqueregions[5], "</b> was the most common industry for those living in ", textprod(data5), " and <b>",
+        uniqueregions[6], "</b> was the most common industry for those living in ", textprod(data6), "."
+      )
+    } else {
+      # Just going to recreate the above with the collapse flag. Leaving it in
+      # the else area for now, but should cover all of the above if statements in one go.
+      print("### Using new code for region text...")
```
Comment (cjrace) [inline R/subject_industry.R]: Another print()?

**Review by cjrace — CHANGES_REQUESTED** [review]
Couple of rogue print statements and also some errors still hanging around on the tables (both for proportions and median earnings:

e.g. Sex > 5 YAG > Level 7 (research) > Architecture

image

image

e.g. Sex > 5 YAG > Level 8 > Architecture

image

image

e.g. Sex > 10 YAG > Level 7 (research) > Architecture

image

image

File: R/subject_industry.R (hunk 1 above)
Comment (rmbielby) [reply R/subject_industry.R]: Guessing I must have been at some point, will delete it.

File: R/subject_industry.R (hunk 2 above)
Comment (rmbielby) [reply R/subject_industry.R]: Yep, that ones definitely me, checking I knew for definite how the if statements were being negotiated. Will delete that too!

**Comment by cjrace** [comment]
Another weird error, don't know if some of this is just around handling when the data isn't there as it's filtered at too fine a level?

Sex > 10 YAG > Level 7 (research) > PGCE

image
image

**Comment by chfoster** [comment]
I think the PGCE issue will be because it should only appear as an option for level 7 (taught). The same goes for MBA

**Review by cjrace — APPROVED** [review]

**Pull request merged**